aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp

```diff
@@ -3012,82 +3018,173 @@ static void apply_lu_solve_batched_magma(const Tensor& b, const Tensor& lu, cons
+  // There is a bug in MAGMA when TransposeType is transpose or conj-transpose.
+  TORCH_INTERNAL_ASSERT(trans == TransposeType::NoTranspose);
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(LU.scalar_type(), "lu_solve_batched_magma", [&]{
+    apply_lu_solve_batched_magma<scalar_t>(LU, pivots, B, trans);
   });
 }
 
-static void lu_solve_looped_magma(const Tensor& b, const Tensor& lu, const Tensor& pivots, TransposeType trans) {
-  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(b.scalar_type(), "lu_solve_looped_magma", [&]{
-    apply_lu_solve_looped_magma<scalar_t>(b, lu, pivots, trans);
+static void lu_solve_looped_magma(const Tensor& LU, const Tensor& pivots, const Tensor& B, TransposeType trans) {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(LU.scalar_type(), "lu_solve_looped_magma", [&]{
+    apply_lu_solve_looped_magma<scalar_t>(LU, pivots, B, trans);
   });
 }
 
-namespace {
-
-c10::MaybeOwned<Tensor> maybe_expand_lu(const Tensor& b, const Tensor& lu) {
-  if (batchCount(b) != batchCount(lu)) {
-    IntArrayRef b_batch_size(b.sizes().data(), b.dim() - 2);
-    DimVector expand_size(b_batch_size);
-    expand_size.insert(expand_size.end(), {lu.size(-2), lu.size(-1)});
+c10::MaybeOwned<Tensor> maybe_expand_lu(const Tensor& B, const Tensor& LU) {
+  // B and LU have the same number of dimensions
+  if (batchCount(B) != batchCount(LU)) {
+        auto n = B.dim();
+    auto expand_shape = DimVector(B.sizes().slice(0, n - 2));
+    expand_shape.append({LU.size(-2), LU.size(-1)});
     return c10::MaybeOwned<Tensor>::owned(
-        cloneBatchedColumnMajor(lu.expand(expand_size)));
+        cloneBatchedColumnMajor(LU.expand(expand_shape)));
   } else {
-    return c10::MaybeOwned<Tensor>::borrowed(lu);
+    return c10::MaybeOwned<Tensor>::borrowed(LU);
   }
 }
 
-c10::MaybeOwned<Tensor> maybe_expand_pivots(const Tensor& b,const Tensor& pivots) {
-  if (batchCount(b) != batchCount(pivots.unsqueeze(-1))) {
-    IntArrayRef b_batch_size(b.sizes().data(), b.dim() - 2);
-    DimVector expand_size(b_batch_size);
-    expand_size.insert(expand_size.end(), {pivots.size(-1)});
-    return c10::MaybeOwned<Tensor>::owned(
-        pivots.expand(expand_size).clone(at::MemoryFormat::Contiguous));
+c10::MaybeOwned<Tensor> maybe_expand_pivots(const Tensor& B, const Tensor& pivots) {
+  // B and pivots have the same number of dimensions
+  if (batchCount(B) != batchCount(pivots.unsqueeze(-1))) {
+    auto expand_shape = DimVector(B.sizes().slice(0, B.dim() - 2));
+    expand_shape.push_back(pivots.size(-1));
+    return c10::MaybeOwned<Tensor>::owned(pivots.expand(expand_shape).contiguous());
   } else {
     return c10::MaybeOwned<Tensor>::borrowed(pivots);
   }
 }
 
-}  // anonymous namespace
+static void lu_solve_kernel(const Tensor& LU, const Tensor& pivots, const Tensor& B, TransposeType trans) {
+  // Trivial case. Remove it once `torch.solve` is removed, as linalg.solve already shortcuts this case
```

Comment:
@lezcano is it time to remove this now?

Reply:
It's rather minor, but sure, happy to approve a PR that removes this provided that CI passes